# Preparation

Preparation biasa dilakukan untuk mempersiapkan data sebelum masuk dalam tahap pemodelan. <br>
Berikut adalah tahapan yang akan dilalui pada data `SC_HW1_bank_data.csv` sebelum tahap pemodelan :
1. Import Library
2. Input Dataset
3. Preprocessing
4. Train-Test Split

## Import Library

In [2]:
import pandas as pd
import numpy as np

## Input Dataset

In [3]:
#Membaca data dan memasukkannya ke dalam bentuk Pandas Dataframe
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')
df

RowNumber  CustomerId Geography  Gender  Age  Tenure    Balance  \
0             1    15634602    France  Female   42       2       0.00   
1             2    15647311     Spain  Female   41       1   83807.86   
2             3    15619304    France  Female   42       8  159660.80   
3             4    15701354    France  Female   39       1       0.00   
4             5    15737888     Spain  Female   43       2  125510.82   
...         ...         ...       ...     ...  ...     ...        ...   
9995       9996    15606229    France    Male   39       5       0.00   
9996       9997    15569892    France    Male   35      10   57369.61   
9997       9998    15584532    France  Female   36       7       0.00   
9998       9999    15682355   Germany    Male   42       3   75075.31   
9999      10000    15628319    France  Female   28       4  130142.79   

      NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0                 1          1               1        101348.88       1  
1                 1          0               1        112542.58       0  
2                 3          1               0        113931.57       1  
3                 2          0               0         93826.63       0  
4                 1          1               1         79084.10       0  
...             ...        ...             ...              ...     ...  
9995              2          1               0         96270.64       0  
9996              1          1               1        101699.77       0  
9997              1          0               1         42085.58       1  
9998              2          1               0         92888.52       1  
9999              1          1               0         38190.78       0  

[10000 rows x 12 columns]

In [4]:
#Jalankan code untuk mengecek nama kolom yang tersedia
df.columns

Index(['RowNumber', 'CustomerId', 'Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited'],
      dtype='object')

In [5]:
#Hilangkan kolom yang dirasa tidak relevan dengan model (contoh: ID). None dapat diisi dengan nama-nama kolom yang akan digunakan.
#Contoh df = df[['X1','X2', 'Y']].copy()

#START CODE
df1 = df[['Geography','Gender','Age','Tenure','Balance','EstimatedSalary']].copy()
df1

Geography  Gender  Age  Tenure    Balance  EstimatedSalary
0       France  Female   42       2       0.00        101348.88
1        Spain  Female   41       1   83807.86        112542.58
2       France  Female   42       8  159660.80        113931.57
3       France  Female   39       1       0.00         93826.63
4        Spain  Female   43       2  125510.82         79084.10
...        ...     ...  ...     ...        ...              ...
9995    France    Male   39       5       0.00         96270.64
9996    France    Male   35      10   57369.61        101699.77
9997    France  Female   36       7       0.00         42085.58
9998   Germany    Male   42       3   75075.31         92888.52
9999    France  Female   28       4  130142.79         38190.78

[10000 rows x 6 columns]

## Preprocessing

In [6]:
#Lakukan One-Hot Encoder untuk data categorical, dengan fungsi pandas get_dummies

#START CODE
# df = None

df2 = pd.get_dummies(df1, columns=['Geography','Gender'], drop_first=True)
df2

Age  Tenure    Balance  EstimatedSalary  Geography_Germany  \
0      42       2       0.00        101348.88                  0   
1      41       1   83807.86        112542.58                  0   
2      42       8  159660.80        113931.57                  0   
3      39       1       0.00         93826.63                  0   
4      43       2  125510.82         79084.10                  0   
...   ...     ...        ...              ...                ...   
9995   39       5       0.00         96270.64                  0   
9996   35      10   57369.61        101699.77                  0   
9997   36       7       0.00         42085.58                  0   
9998   42       3   75075.31         92888.52                  1   
9999   28       4  130142.79         38190.78                  0   

      Geography_Spain  Gender_Male  
0                   0            0  
1                   1            0  
2                   0            0  
3                   0            0  
4                   1            0  
...               ...          ...  
9995                0            1  
9996                0            1  
9997                0            0  
9998                0            1  
9999                0            0  

[10000 rows x 7 columns]

In [7]:
#Pisahkan mana X (feature) dengan Y,
#Y adalah kolom "Exited"

#START CODE
X = df["Age"]
y = df["Exited"]

In [8]:
#Lakukan Scaler dan/atau Noermalisasi jika diperlukan
from sklearn.preprocessing import MinMaxScaler

# Inisialisasi scaler
scaler = MinMaxScaler()

# Melakukan scaling pada fitur "EstimatedSalary"
X_transform = scaler.fit_transform(df2[["Age"]])

# Hasil akan berupa array numpy, Anda dapat mengonversinya ke dalam Pandas DataFrame jika diperlukan
X_scaled = pd.DataFrame(X_transform, columns=["Age"])

# Menyimpan data X yang telah di-scaling
df2["Age"] = X_scaled["Age"]

df2

Age  Tenure    Balance  EstimatedSalary  Geography_Germany  \
0     0.324324       2       0.00        101348.88                  0   
1     0.310811       1   83807.86        112542.58                  0   
2     0.324324       8  159660.80        113931.57                  0   
3     0.283784       1       0.00         93826.63                  0   
4     0.337838       2  125510.82         79084.10                  0   
...        ...     ...        ...              ...                ...   
9995  0.283784       5       0.00         96270.64                  0   
9996  0.229730      10   57369.61        101699.77                  0   
9997  0.243243       7       0.00         42085.58                  0   
9998  0.324324       3   75075.31         92888.52                  1   
9999  0.135135       4  130142.79         38190.78                  0   

      Geography_Spain  Gender_Male  
0                   0            0  
1                   1            0  
2                   0            0  
3                   0            0  
4                   1            0  
...               ...          ...  
9995                0            1  
9996                0            1  
9997                0            0  
9998                0            1  
9999                0            0  

[10000 rows x 7 columns]

In [9]:
#Ini digunakan jika dilakukan scaler/Normalisasi. Jika tidak, code ini bisa dilewat dan diganti dengan code yang ada di dalam komen
# X_transform = pd.DataFrame(X_transform, columns = X.columns)
# #X_transform = X.copy()

# Melakukan scaling pada fitur "EstimatedSalary"
X_transform = scaler.fit_transform(df2[["Age"]])

# Menggantikan nilai X dengan X yang sudah di-transformasi
X_transform = pd.DataFrame(X_transform, columns=["Age"])


## Train-Test Split

In [10]:
#Split menjadi train dan test dengan test_size 25%
#Tidak perlu mengubah code ini

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_transform,y,test_size = 0.25,random_state = 123)

# Modeling

## Model1

### Soal :
Jelaskan secara Singkat Model pertama yang digunakan!

In [74]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix

# Split menjadi train dan test dengan test_size 25%
# Tidak perlu mengubah code ini
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_transform, y, test_size=0.25, random_state=123)

# Membuat objek model Multinomial Naive Bayes
model = MultinomialNB()

# Membuat daftar hyperparameter yang akan diuji
param_grid = {
    'alpha': [0.1, 0.5, 1.0, 2.0, 5.0],  # Hyperparameter alpha yang akan diuji
}

# Membuat objek GridSearchCV untuk mencari parameter terbaik
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=5)

# Melatih model pada data pelatihan dengan menggunakan GridSearchCV
grid_search.fit(X_train, y_train)

# Melihat parameter terbaik yang telah ditemukan
best_params = grid_search.best_params_
print("Parameter terbaik:", best_params)

# Membuat model final dengan parameter terbaik yang ditemukan
best_model = MultinomialNB(alpha=best_params['alpha'])
best_model.fit(X_train, y_train)

# Mengevaluasi model pada data pengujian
y_pred = best_model.predict(X_test)

# Menampilkan classification report dengan penanganan zero division
print("Classification Report:")
print(classification_report(y_test, y_pred, zero_division=0))

# Menampilkan confusion matrix
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

# Menampilkan akurasi model
accuracy = accuracy_score(y_test, y_pred)
print("Akurasi model di atas data pengujian:", accuracy)

Parameter terbaik: {'alpha': 0.1}
Classification Report:
              precision    recall  f1-score   support

           0       0.79      1.00      0.88      1983
           1       0.00      0.00      0.00       517

    accuracy                           0.79      2500
   macro avg       0.40      0.50      0.44      2500
weighted avg       0.63      0.79      0.70      2500

Confusion Matrix:
[[1983    0]
 [ 517    0]]
Akurasi model di atas data pengujian: 0.7932


## Model2
### Soal :
Jelaskan secara Singkat Model ke-2 yang digunakan!

In [13]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix

# Membuat objek model Decision Tree
decision_tree_model = DecisionTreeClassifier()

# Membuat daftar hyperparameter yang akan diuji
param_grid = {
    'criterion': ['gini', 'entropy'],  # Kriteria pemilihan node
    'max_depth': [None, 10, 20, 30, 40],  # Maksimum kedalaman pohon
    'min_samples_split': [2, 5, 10],  # Jumlah sampel minimum untuk membagi node internal
    'min_samples_leaf': [1, 2, 4]  # Jumlah sampel minimum pada node daun
}

# Membuat objek GridSearchCV untuk mencari parameter terbaik
tree_grid_search = GridSearchCV(estimator=decision_tree_model, param_grid=param_grid, cv=5)

# Melatih model pada data pelatihan dengan menggunakan GridSearchCV
tree_grid_search.fit(X_train, y_train)

# Melihat parameter terbaik yang telah ditemukan
best_tree_params = tree_grid_search.best_params_
print("Parameter terbaik untuk Decision Tree:", best_tree_params)

# Membuat model Decision Tree dengan parameter terbaik yang ditemukan
best_tree_model = DecisionTreeClassifier(criterion=best_tree_params['criterion'],
                                         max_depth=best_tree_params['max_depth'],
                                         min_samples_split=best_tree_params['min_samples_split'],
                                         min_samples_leaf=best_tree_params['min_samples_leaf'])
best_tree_model.fit(X_train, y_train)

# Mengevaluasi model pada data pengujian
y_tree_pred = best_tree_model.predict(X_test)

# Menampilkan classification report dengan penanganan zero division
print("Classification Report untuk Decision Tree:")
print(classification_report(y_test, y_tree_pred, zero_division=0))

# Menampilkan confusion matrix
print("Confusion Matrix untuk Decision Tree:")
print(confusion_matrix(y_test, y_tree_pred))

# Menampilkan akurasi model Decision Tree
tree_accuracy = accuracy_score(y_test, y_tree_pred)
print("Akurasi model Decision Tree di atas data pengujian:", tree_accuracy)


Parameter terbaik untuk Decision Tree: {'criterion': 'gini', 'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 5}
Classification Report untuk Decision Tree:
              precision    recall  f1-score   support

           0       0.83      0.95      0.88      1983
           1       0.53      0.24      0.33       517

    accuracy                           0.80      2500
   macro avg       0.68      0.59      0.61      2500
weighted avg       0.77      0.80      0.77      2500

Confusion Matrix untuk Decision Tree:
[[1874  109]
 [ 392  125]]
Akurasi model Decision Tree di atas data pengujian: 0.7996


## Model3
### Soal :
Jelaskan secara Singkat Model ke-3 yang digunakan!

In [10]:
# #Pilih salah satu metode Machine Leaarning

from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

# Inisialisasi model Random Forest
model_rf = RandomForestClassifier()

# Daftar hyperparameter yang akan diuji
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [10, 20, 30, None],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

# Inisialisasi GridSearchCV
grid = GridSearchCV(estimator=model_rf, param_grid=param_grid, scoring='accuracy', cv=10)

# Melatih model pada berbagai kombinasi hyperparameter
grid.fit(X_train, y_train)

# Mendapatkan hyperparameter terbaik
best_params = grid.best_params_

print("Hyperparameter terbaik:", best_params)

# Membuat model Random Forest dengan hyperparameter terbaik
best_model_rf = RandomForestClassifier(**best_params)

# Melatih model dengan hyperparameter terbaik
best_model_rf.fit(X_train, y_train)

# Melakukan prediksi pada data uji
y_pred_rf = best_model_rf.predict(X_test)

# Evaluasi model
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

accuracy_rf = accuracy_score(y_test, y_pred_rf)
conf_matrix_rf = confusion_matrix(y_test, y_pred_rf)
class_report_rf = classification_report(y_test, y_pred_rf)

print("Akurasi Random Forest:", accuracy_rf)
print("\nConfusion Matrix Random Forest:\n", conf_matrix_rf)
print("\nClassification Report Random Forest:\n", class_report_rf)

Hyperparameter terbaik: {'max_depth': 10, 'min_samples_leaf': 1, 'min_samples_split': 10, 'n_estimators': 100}
Akurasi Random Forest: 0.7932

Confusion Matrix Random Forest:
 [[1980    3]
 [ 514    3]]

Classification Report Random Forest:
               precision    recall  f1-score   support

           0       0.79      1.00      0.88      1983
           1       0.50      0.01      0.01       517

    accuracy                           0.79      2500
   macro avg       0.65      0.50      0.45      2500
weighted avg       0.73      0.79      0.70      2500



## Tarik Kesimpulan Model Mana yang terbaik beserta alasannya

dari ketiga model algoritma diatas, model terbaik didapatkan oleh decision tree, dimana akurasi yang didapatkan adalah sebesar 0.7996 atau 80%.
hal ini bisa didapatkan karena terdapat GridSearchCV untuk mencari parameter terbaik dalam pengujian